### PR TITLE
Add insert_all

### PR DESCRIPTION
**Pull request by ruslandoga**
This PR adds a helper NIF that can be used to bulk-insert "unlimited" number of rows.

Should be enough to clоse https://github.com/elixir-sqlite/exqlite/issues/129 once it gets into `ecto_sqlite3`

Regarding [performance:](https://github.com/elixir-sqlite/exqlite/issues/129#issuecomment-937400559) XQLite uses this approach in `insert_all` benchmark in https://github.com/ruslandoga/elixir-sqlite-bench

### CODE REVIEW

File: c_src/sqlite3_nif.c [hunk 1]
```diff
@@ -691,6 +691,102 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+    // process rows
+    while (enif_get_list_cell(env, rows, &head, &tail)) {
+        sqlite3_reset(statement->statement);
+
+        // bind row
+        for (unsigned int i = 1; i <= stmt_param_count; i++) {
+            ERL_NIF_TERM param;
+
+            if (!enif_get_list_cell(env, head, &param, &head))
+                return raise_badarg(env, param);
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:732]:
```suggestion
            if (!enif_get_list_cell(env, head, &param, &head)) {
                return raise_badarg(env, param);
            }
```

File: c_src/sqlite3_nif.c [hunk 2]
```diff
@@ -691,6 +691,102 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+
+            if (enif_is_identical(param, am_nil)) {
+                rc = sqlite3_bind_null(statement->statement, i);
+            } else {
+                switch (types_array[i - 1]) {
+                    case SQLITE_INTEGER: {
+                        ErlNifSInt64 i64;
+                        if (!enif_get_int64(env, param, &i64))
+                            return raise_badarg(env, param);
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:741]:
```suggestion
                        if (!enif_get_int64(env, param, &i64)) {
                            return raise_badarg(env, param);
                        }
```

File: c_src/sqlite3_nif.c [hunk 3]
```diff
@@ -691,6 +691,102 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+
+                        rc = sqlite3_bind_int64(statement->statement, i, i64);
+                        break;
+                    }
+
+                    case SQLITE_FLOAT: {
+                        double f64;
+                        if (!enif_get_double(env, param, &f64))
+                            return raise_badarg(env, param);
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:750]:
```suggestion
                        if (!enif_get_double(env, param, &f64)) {
                            return raise_badarg(env, param);
                        }
```

File: c_src/sqlite3_nif.c [hunk 4]
```diff
@@ -691,6 +691,102 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+
+                        rc = sqlite3_bind_double(statement->statement, i, f64);
+                        break;
+                    }
+
+                    case SQLITE_TEXT: {
+                        ErlNifBinary text;
+                        if (!enif_inspect_binary(env, param, &text))
+                            return raise_badarg(env, param);
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:759]:
```suggestion
                        if (!enif_inspect_binary(env, param, &text)) {
                            return raise_badarg(env, param);
                        }
```

File: c_src/sqlite3_nif.c [hunk 5]
```diff
@@ -691,6 +691,102 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+
+                        rc = sqlite3_bind_text(statement->statement, i, (char*)text.data, text.size, SQLITE_TRANSIENT);
+                        break;
+                    }
+
+                    case SQLITE_BLOB: {
+                        ErlNifBinary blob;
+                        if (!enif_inspect_binary(env, param, &blob))
+                            return raise_badarg(env, param);
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:768]:
```suggestion
                        if (!enif_inspect_binary(env, param, &blob)) {
                            return raise_badarg(env, param);
                        }
```

File: c_src/sqlite3_nif.c [hunk 6]
```diff
@@ -691,6 +691,102 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+
+                        rc = sqlite3_bind_blob(statement->statement, i, (char*)blob.data, blob.size, SQLITE_TRANSIENT);
+                        break;
+                    }
+                }
+            }
+
+            if (rc != SQLITE_OK)
+                return make_sqlite3_error_tuple(env, rc, sqlite3_db_handle(statement->statement));
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:777]:
```suggestion
            if (rc != SQLITE_OK) {
                return make_sqlite3_error_tuple(env, rc, sqlite3_db_handle(statement->statement));
            }
```

File: c_src/sqlite3_nif.c [hunk 7]
```diff
@@ -691,6 +691,102 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+    int types_array[stmt_param_count];
+
+    ERL_NIF_TERM types = argv[1];
+    ERL_NIF_TERM rows  = argv[2];
+    ERL_NIF_TERM head, tail;
+
+    // process types
+    for (unsigned int i = 0; i < stmt_param_count; i++) {
+        if (!enif_get_list_cell(env, types, &head, &tail))
+            return raise_badarg(env, types);
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:711]:
```suggestion
        if (!enif_get_list_cell(env, types, &head, &tail)) {
            return raise_badarg(env, types);
        }
```

File: lib/exqlite/sqlite3.ex [hunk 8]
```diff
@@ -277,6 +277,42 @@ defmodule Exqlite.Sqlite3 do
     fetch_all(conn, statement, chunk_size)
   end
 
+  @doc """
+  Bulk-inserts rows into a prepared statement. Must be called inside a transaction.
+
```
Comment (warmwaffles) [inline lib/exqlite/sqlite3.ex:282]:
```suggestion
  @doc """
  Bulk-inserts rows into a prepared statement. Must be called inside a transaction.

  The `types` parameter is type of each column in the order they are specified 
  in the parameter `rows`. For example, `[["a", 1, 1.0]]` would have `types`
  specified as `[:text, :integer, :float]`.
```

File: c_src/sqlite3_nif.c [hunk 9]
```diff
@@ -691,6 +691,112 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+
+    int stmt_param_count = sqlite3_bind_parameter_count(statement->statement);
+    int types_array[stmt_param_count];
+
+    ERL_NIF_TERM types = argv[1];
+    ERL_NIF_TERM rows  = argv[2];
+    ERL_NIF_TERM head, tail;
+
+    // process types
+    for (unsigned int i = 0; i < stmt_param_count; i++) {
```
Comment (ruslandoga) [inline c_src/sqlite3_nif.c:710]:
```suggestion
    for (int i = 0; i < stmt_param_count; i++) {
```

File: lib/exqlite/sqlite3.ex [hunk 10]
```diff
@@ -277,6 +277,45 @@ defmodule Exqlite.Sqlite3 do
...
+  The `types` parameter is type of each column in the order they are specified 
+  in the parameter `rows`. For example, `[["a", 1, 1.0]]` would have `types`
+  specified as `[:text, :integer, :float]`.
+      iex> {:ok, conn} = Sqlite3.open(":memory:", [:readonly])
```
Comment (ruslandoga) [inline lib/exqlite/sqlite3.ex:286]:
```suggestion

      iex> {:ok, conn} = Sqlite3.open(":memory:", [:readonly])
```

File: c_src/sqlite3_nif.c [hunk 11]
```diff
@@ -691,6 +691,112 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+    }
+
+    int rc;
+
+    // process rows
+    while (enif_get_list_cell(env, rows, &head, &tail)) {
+        sqlite3_reset(statement->statement);
+
+        // bind row
+        for (unsigned int i = 1; i <= stmt_param_count; i++) {
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:731]:
```suggestion
        for (int i = 1; i <= stmt_param_count; i++) {
```

**Comment by warmwaffles** [comment]
This looks good.

**Comment by warmwaffles** [comment]
I do not like how often the windows test flaps.

**Comment by warmwaffles** [comment]
Ugh

```
c_src\sqlite3_nif.c(703): error C2057: expected constant expression
c_src\sqlite3_nif.c(703): error C2466: cannot allocate an array of constant size 0
c_src\sqlite3_nif.c(703): error C2133: 'types_array': unknown size
```

File: c_src/sqlite3_nif.c [hunk 12]
```diff
@@ -691,6 +691,112 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+static ERL_NIF_TERM
+exqlite_insert_all(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
+{
+    statement_t* statement;
+    if (!enif_get_resource(env, argv[0], statement_type, (void**)&statement)) {
+        return raise_badarg(env, argv[0]);
+    }
+
+    int stmt_param_count = sqlite3_bind_parameter_count(statement->statement);
+    int types_array[stmt_param_count];
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:703]: This needs to be dynamically allocated. It won't work with the windows compiler.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c:703]: Ah, I see. I'll fix it properly tomorrow!

Comment (warmwaffles) [reply c_src/sqlite3_nif.c:703]: I wonder if the Zig compiler would be a better long term option for Windows. 🤔 I'll have to play with that idea later.

Comment (warmwaffles) [reply c_src/sqlite3_nif.c:703]: @ruslandoga this is the last bit that we need to update. Otherwise this looks good to me.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c:703]: 👋 @warmwaffles 

Right, sorry. Got a bit distracted again. I'll finish it up this weekend!

Comment (ruslandoga) [reply c_src/sqlite3_nif.c]: I pushed a commit but I don't like it ...

> I wonder if the Zig compiler would be a better long term option for Windows.

Maybe just Clang? It seems to be supported on Windows, and can be use for pre-compilation. We can also document that Windows users, if they want to compile the library from scratch, should use Clang.

It seems like too much effort trying to adapt to different compilers :)

Comment (ruslandoga) [reply c_src/sqlite3_nif.c]: Side issue: there seems to be something wrong with Makefile, it doesn't recompile on sqlite_nif.c changes.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c]: Seems to be `cc_precompiler` issue, disabling it makes `mix compile` recompile touched files.

File: c_src/sqlite3_nif.c [hunk 13]
```diff
@@ -700,7 +700,14 @@ exqlite_insert_all(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
     }
 
     int stmt_param_count = sqlite3_bind_parameter_count(statement->statement);
-    int types_array[stmt_param_count];
+
+    // TODO: MSVC can't do VLA
+    // int types_array[stmt_param_count];
+    int* types_array = (int*)malloc(stmt_param_count * sizeof(int));
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c]: Use `exqlite_malloc` for this. This way the VM will know about the memory allocated and will count it correctly. Also `exqlite_free` to free the memory.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c]: Right, I did it in the very next commit: https://github.com/elixir-sqlite/exqlite/pull/306/commits/cf69cd54cec3d111b7e648deac36f028d04ad227

Comment (warmwaffles) [reply c_src/sqlite3_nif.c]: ah I was walking the changes.

File: c_src/sqlite3_nif.c [hunk 14]
```diff
@@ -691,6 +691,120 @@ exqlite_step(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
...
+            }
+
+            if (rc != SQLITE_OK) {
+                return make_sqlite3_error_tuple(env, rc, sqlite3_db_handle(statement->statement));
+            }
+        }
+
+        rc = sqlite3_step(statement->statement);
+        if (rc != SQLITE_DONE) {
+            return make_sqlite3_error_tuple(env, rc, sqlite3_db_handle(statement->statement));
```
Comment (warmwaffles) [inline c_src/sqlite3_nif.c:798]: We are missing cleanup code here. We need to `enif_free(types_array)` before all of these error returns.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c:798]: Ah, right.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c:798]: I think I'll need to find another way to pass the types information. The current approach is getting complicated.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c:798]: Sorry!

Comment (ruslandoga) [reply c_src/sqlite3_nif.c:798]: > This is just the nature of C.

I think it should be possible to implement `insert_all` or a similar function for bulk inserts with no manual memory allocation. It's only used in a small part of the function -- just to know what types to insert -- but ends up affecting all control flow.

Comment (ruslandoga) [reply c_src/sqlite3_nif.c:798]: All `Repo.insert_all` calls would go through this NIF and the performance improvement is quite significant (seems close to x2 from my benchmarks, but potentially higher becuase it allows reusing a single prepared statement for all inserts into a single table) so the I think it's worth it.

**Pull request closed without merging**